File: runtimes/java/src/eet/Vm.java
package eet;

import java.io.IOException;
import java.io.OutputStream;
import java.util.Arrays;

/**
 * The EET interpreter (spec section 5).
 *
 * <p>The loop is deliberately literal: fetch, bounds-check the immediates, dispatch. Where
 * Java's native behaviour already is the specified behaviour it is used directly and the
 * reason is named in a comment. It differs in exactly three places, each corrected and
 * cited below: {@code u32} branch targets, which are not {@code int}; the {@code addr + len}
 * of {@code prints}, which can overflow one; and {@code byte}, which Java signs.
 */
final class Vm {

    /** One activation record: private locals, a private operand stack, a return address. */
    private static final class Frame {

        /**
         * Frames are cheap and plentiful and almost every one uses a handful of slots, so
         * the operand stack starts small and grows to the ceiling of section 1 on demand
         * rather than costing 4KB per call.
         */
        private static final int INITIAL_STACK = 16;

        final int[] locals;
        final int returnPc;
        int[] stack = new int[INITIAL_STACK];
        int sp;

        Frame(int nlocals, int returnPc) {
            // Zero-initialised by Java, which is what sections 5.1 and 5.4 ask for.
            this.locals = new int[nlocals];
            this.returnPc = returnPc;
        }
    }

    private final byte[] code;
    private final byte[] data;
    private final int codeLen;
    private final int[] globals;
    private final OutputStream stdout;

    private final Frame[] frames = new Frame[Isa.MAX_CALL_DEPTH];
    private int depth;

    private int pc;

    /** First byte of the instruction being executed; every trap reports it (section 6). */
    private int opPc;

    private Vm(EetModule module, OutputStream stdout) {
        this.code = module.code();
        this.data = module.data();
        this.codeLen = code.length;
        this.globals = new int[module.nglobals()];
        this.stdout = stdout;
        // The entry frame's return address is never read: it is the sentinel "exit" of
        // section 5.1, and ret recognises the entry frame by the call stack emptying.
        this.frames[0] = new Frame(module.entryLocals(), -1);
        this.depth = 1;
        this.pc = module.entry();
    }

    /**
     * Run {@code module} to completion and return the process exit status (section 5.3).
     *
     * <p>The trap protocol lives here because its ordering is observable: whatever the
     * program already wrote must reach stdout before the trap line reaches stderr.
     */
    static int execute(EetModule module, OutputStream stdout, OutputStream stderr)
            throws IOException {
        Vm vm = new Vm(module, stdout);
        try {
            int status = vm.run();
            stdout.flush();
            return status;
        } catch (Trap trap) {
            stdout.flush();
            stderr.write(trap.line());
            stderr.flush();
            return Isa.EXIT_TRAP;
        }
    }

    private int run() throws IOException {
        while (true) {
            if (pc >= codeLen) {
                // Falling off the end of the code section (section 5.4). No instruction
                // started here, so the address reported is the fall-through pc itself.
                throw new Trap(Trap.Id.T09, pc);
            }
            opPc = pc;
            Isa.Opcode op = Isa.decode(code[pc]);
            pc++;
            if (op == null) {
                throw trap(Trap.Id.T05);
            }
            // One check covers the whole instruction: reading any immediate past the end of
            // code[] is T05 (sections 3 and 5.2). Every case below then reads its own
            // immediates knowing they are there.
            if (pc + op.immediateBytes() > codeLen) {
                throw trap(Trap.Id.T05);
            }

            switch (op) {
                // -- 4.1 stack and control
                case HALT -> {
                    return Isa.EXIT_OK;
                }
                case NOP -> {
                }
                case PUSH -> push(i32());
                case POP -> pop();
                case DUP -> {
                    int a = pop();
                    push(a);
                    push(a);
                }
                case SWAP -> {
                    int b = pop();
                    int a = pop();
                    push(b);
                    push(a);
                }
                case OVER -> {
                    int b = pop();
                    int a = pop();
                    push(a);
                    push(b);
                    push(a);
                }
                case ROT -> {
                    int c = pop();
                    int b = pop();
                    int a = pop();
                    push(b);
                    push(c);
                    push(a);
                }

                // -- 4.2 arithmetic and bitwise, 4.3 comparison
                case NEG -> push(-pop());
                case NOT -> push(~pop());
                case ADD, SUB, MUL, DIV, MOD, AND, OR, XOR, SHL, SHR, USHR,
                        EQ, NE, LT, LE, GT, GE -> {
                    // The top of the stack is the right-hand operand: a b -> a op b.
                    int b = pop();
                    int a = pop();
                    push(apply(op, a, b));
                }

                // -- 4.4 branching and calls
                case JMP -> jump(u32());
                case JZ -> {
                    long target = u32();
                    if (pop() == 0) {
                        jump(target);
                    }
                }
                case JNZ -> {
                    long target = u32();
                    if (pop() != 0) {
                        jump(target);
                    }
                }
                case CALL -> call();
                case RET -> {
                    int value = pop();
                    Frame finished = frames[--depth];
                    if (depth == 0) {
                        // Returning from the entry frame ends the program (section 5.3).
                        return value & 0xFF;
                    }
                    pc = finished.returnPc;
                    push(value);
                }

                // -- 4.5 memory
                case LOAD -> {
                    int idx = u8();
                    int[] locals = frame().locals;
                    if (idx >= locals.length) {
                        throw trap(Trap.Id.T06);
                    }
                    push(locals[idx]);
                }
                case STORE -> {
                    int idx = u8();
                    int[] locals = frame().locals;
                    // The index is validated before the value is popped, so a store that is
                    // both out of range and starved reports T06 and not T01.
                    if (idx >= locals.length) {
                        throw trap(Trap.Id.T06);
                    }
                    locals[idx] = pop();
                }
                case GLOAD -> {
                    int idx = u16();
                    if (idx >= globals.length) {
                        throw trap(Trap.Id.T07);
                    }
                    push(globals[idx]);
                }
                case GSTORE -> {
                    int idx = u16();
                    if (idx >= globals.length) {
                        throw trap(Trap.Id.T07);
                    }
                    globals[idx] = pop();
                }
                case DLOAD -> {
                    int addr = pop();
                    if (addr < 0 || addr >= data.length) {
                        throw trap(Trap.Id.T08);
                    }
                    // Java's byte is signed; section 4.5 wants the value zero-extended into
                    // 0..255, so a bare data[addr] would be wrong for every byte over 127.
                    push(data[addr] & 0xFF);
                }
                case GLOADX -> {
                    int idx = pop();
                    if (idx < 0 || idx >= globals.length) {
                        throw trap(Trap.Id.T07);
                    }
                    push(globals[idx]);
                }
                case GSTOREX -> {
                    // The index sits on top of the value: v idx -> (section 4.5).
                    int idx = pop();
                    int value = pop();
                    if (idx < 0 || idx >= globals.length) {
                        throw trap(Trap.Id.T07);
                    }
                    globals[idx] = value;
                }

                // -- 4.6 output
                case PRINT -> stdout.write(Ops.decimal(pop()));
                case PRINTC -> stdout.write(pop() & 0xFF);
                case PRINTS -> {
                    int len = pop();
                    int addr = pop();
                    // addr + len is widened before the comparison: both are i32 and their
                    // sum can overflow into a negative int, which would slip past the range
                    // check and hand a bad slice to write() (section 4.6).
                    if (len < 0 || addr < 0 || (long) addr + len > data.length) {
                        throw trap(Trap.Id.T08);
                    }
                    stdout.write(data, addr, len);
                }

                // -- 4.7 diagnostics
                case TRAP -> throw new Trap(
                        Trap.Id.T10, opPc, "trap instruction (code=" + u8() + ")");

                // Isa.Opcode is the single source of truth for what exists; this fires only
                // if a constant is added there and no case is added here.
                default -> throw new AssertionError("unhandled opcode " + op);
            }
        }
    }

    /**
     * The binary operators of sections 4.2 and 4.3, as a table of pure functions.
     *
     * <p>Every one of them is a bare Java operator. That is not laziness: {@link Ops}
     * records the JLS clause behind each, covering the wrapping, the truncation toward
     * zero, the five-bit shift mask and the signedness of the comparisons.
     */
    private int apply(Isa.Opcode op, int a, int b) {
        return switch (op) {
            case ADD -> a + b;
            case SUB -> a - b;
            case MUL -> a * b;
            case DIV -> {
                if (b == 0) {
                    throw trap(Trap.Id.T04);
                }
                yield Ops.div(a, b);
            }
            case MOD -> {
                if (b == 0) {
                    throw trap(Trap.Id.T04);
                }
                yield Ops.mod(a, b);
            }
            case AND -> a & b;
            case OR -> a | b;
            case XOR -> a ^ b;
            case SHL -> a << b;
            case SHR -> a >> b;
            case USHR -> a >>> b;
            case EQ -> a == b ? 1 : 0;
            case NE -> a != b ? 1 : 0;
            case LT -> a < b ? 1 : 0;
            case LE -> a <= b ? 1 : 0;
            case GT -> a > b ? 1 : 0;
            case GE -> a >= b ? 1 : 0;
            // Guards the case-label group in run() against drifting out of step with this
            // table; no opcode outside that group can reach here.
            default -> throw new AssertionError("not a binary operator: " + op);
        };
    }

    /** {@code call target, nargs, nlocals} (spec section 5.4). */
    private void call() {
        long target = u32();
        int nargs = u8();
        int nlocals = u8();

        if (nargs > nlocals) {
            throw trap(Trap.Id.T06);
        }
        if (depth == Isa.MAX_CALL_DEPTH) {
            throw trap(Trap.Id.T03);
        }

        // The return address is the pc after the whole instruction, immediates included.
        Frame callee = new Frame(nlocals, pc);
        // Popped back to front, so the first value the caller pushed becomes locals[0].
        for (int i = nargs - 1; i >= 0; i--) {
            callee.locals[i] = pop();
        }
        if (target >= codeLen) {
            throw trap(Trap.Id.T09);
        }

        frames[depth++] = callee;
        pc = (int) target;
    }

    /**
     * Take a branch, checked at the moment it is taken (section 4.4).
     *
     * <p>The target arrives as a {@code long} because it is encoded {@code u32}: a target at
     * or above 2^31 would be a negative {@code int}, would pass a signed range check, and
     * would then index outside {@code code[]} instead of raising {@code T09}.
     */
    private void jump(long target) {
        if (target >= codeLen) {
            throw trap(Trap.Id.T09);
        }
        pc = (int) target;
    }

    private Frame frame() {
        return frames[depth - 1];
    }

    private void push(int value) {
        Frame frame = frame();
        if (frame.sp == frame.stack.length) {
            grow(frame);
        }
        frame.stack[frame.sp++] = value;
    }

    private int pop() {
        Frame frame = frame();
        if (frame.sp == 0) {
            throw trap(Trap.Id.T01);
        }
        return frame.stack[--frame.sp];
    }

    private void grow(Frame frame) {
        if (frame.sp == Isa.MAX_OPERAND_STACK) {
            throw trap(Trap.Id.T02);
        }
        frame.stack = Arrays.copyOf(
                frame.stack, Math.min(frame.stack.length * 2, Isa.MAX_OPERAND_STACK));
    }

    // -- immediate decoding. All multi-byte integers are little-endian (section 2).

    private int u8() {
        return code[pc++] & 0xFF;
    }

    private int u16() {
        int value = (code[pc] & 0xFF) | ((code[pc + 1] & 0xFF) << 8);
        pc += 2;
        return value;
    }

    private int i32() {
        int value = (code[pc] & 0xFF)
                | ((code[pc + 1] & 0xFF) << 8)
                | ((code[pc + 2] & 0xFF) << 16)
                | ((code[pc + 3] & 0xFF) << 24);
        pc += 4;
        return value;
    }

    /** The same four bytes as {@link #i32()}, read unsigned; see {@link #jump}. */
    private long u32() {
        return i32() & 0xFFFF_FFFFL;
    }

    private Trap trap(Trap.Id id) {
        return new Trap(id, opPc);
    }
}

File: runtimes/python/eetvm/vm.py
"""The EET interpreter -- the reference implementation (spec section 5).

Correctness beats speed here on purpose. Where a faster shape would obscure the
correspondence with the specification, the slower and more literal shape wins, because
four other runtimes are read against this one.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from typing import BinaryIO, List

from . import isa, ops
from .binary import Module


class Trap(Exception):
    """A deterministic runtime fault (spec section 6)."""

    def __init__(self, trap_id: str, pc: int, detail: str | None = None) -> None:
        super().__init__(trap_id)
        self.trap_id = trap_id
        self.pc = pc
        self.detail = detail

    def line(self) -> bytes:
        return isa.trap_line(self.trap_id, self.pc, self.detail)


@dataclass
class Frame:
    """One activation record: private locals, private operand stack, return address."""

    locals: List[int]
    return_pc: int
    stack: List[int] = field(default_factory=list)


class VM:
    """A single EET machine bound to one output stream."""

    def __init__(self, module: Module, stdout: BinaryIO) -> None:
        self.module = module
        self.code = module.code
        self.data = module.data
        self.code_len = len(module.code)
        self.stdout = stdout
        self.globals: List[int] = [0] * module.nglobals
        self.frames: List[Frame] = [Frame([0] * module.entry_locals, -1)]
        self.pc = module.entry
        #: Address of the instruction currently executing; every trap reports this.
        self.op_pc = module.entry

    # -- frame and stack helpers --------------------------------------------------------

    @property
    def frame(self) -> Frame:
        return self.frames[-1]

    def push(self, value: int) -> None:
        stack = self.frames[-1].stack
        if len(stack) >= isa.MAX_OPERAND_STACK:
            raise Trap("T02", self.op_pc)
        stack.append(value)

    def pop(self) -> int:
        stack = self.frames[-1].stack
        if not stack:
            raise Trap("T01", self.op_pc)
        return stack.pop()

    # -- instruction decoding -----------------------------------------------------------

    def _fetch(self, fmt: str, size: int) -> int:
        if self.pc + size > self.code_len:
            raise Trap("T05", self.op_pc)
        (value,) = struct.unpack_from(fmt, self.code, self.pc)
        self.pc += size
        return value

    def _u8(self) -> int:
        return self._fetch("<B", 1)

    def _u16(self) -> int:
        return self._fetch("<H", 2)

    def _u32(self) -> int:
        return self._fetch("<I", 4)

    def _i32(self) -> int:
        return self._fetch("<i", 4)

    def _jump(self, target: int) -> None:
        if target >= self.code_len:
            raise Trap("T09", self.op_pc)
        self.pc = target

    # -- the main loop ------------------------------------------------------------------

    def run(self) -> int:
        """Execute until termination. Returns the process exit status.

        Raises :class:`Trap`; the caller is responsible for flushing stdout and printing
        the trap line, because the spec requires prior output to survive a trap.
        """
        while True:
            if self.pc >= self.code_len:
                # Falling off the end of the code section, spec section 5.4.
                raise Trap("T09", self.pc)
            self.op_pc = self.pc
            op = self.code[self.pc]
            self.pc += 1

            result = self._execute(op)
            if result is not None:
                return result

    def _execute(self, op: int) -> int | None:
        """Run one instruction. Returns an exit status only when the program terminates."""
        # -- 4.1 stack and control
        if op == 0x00:  # halt
            return isa.EXIT_OK
        if op == 0x01:  # nop
            return None
        if op == 0x02:  # push
            self.push(self._i32())
            return None
        if op == 0x03:  # pop
            self.pop()
            return None
        if op == 0x04:  # dup
            a = self.pop()
            self.push(a)
            self.push(a)
            return None
        if op == 0x05:  # swap
            b, a = self.pop(), self.pop()
            self.push(b)
            self.push(a)
            return None
        if op == 0x06:  # over: a b -> a b a
            b, a = self.pop(), self.pop()
            self.push(a)
            self.push(b)
            self.push(a)
            return None
        if op == 0x07:  # rot: a b c -> b c a
            c, b, a = self.pop(), self.pop(), self.pop()
            self.push(b)
            self.push(c)
            self.push(a)
            return None

        # -- 4.2 arithmetic and bitwise
        if 0x10 <= op <= 0x1C:
            return self._arith(op)

        # -- 4.3 comparison
        if 0x20 <= op <= 0x25:
            b, a = self.pop(), self.pop()
            self.push(1 if _COMPARE[op](a, b) else 0)
            return None

        # -- 4.4 branching and calls
        if op == 0x30:  # jmp
            self._jump(self._u32())
            return None
        if op == 0x31:  # jz
            target = self._u32()
            if self.pop() == 0:
                self._jump(target)
            return None
        if op == 0x32:  # jnz
            target = self._u32()
            if self.pop() != 0:
                self._jump(target)
            return None
        if op == 0x33:  # call
            return self._call()
        if op == 0x34:  # ret
            return self._ret()

        # -- 4.5 memory
        if op == 0x40:  # load
            idx = self._u8()
            slots = self.frame.locals
            if idx >= len(slots):
                raise Trap("T06", self.op_pc)
            self.push(slots[idx])
            return None
        if op == 0x41:  # store
            idx = self._u8()
            slots = self.frame.locals
            if idx >= len(slots):
                raise Trap("T06", self.op_pc)
            slots[idx] = self.pop()
            return None
        if op == 0x42:  # gload
            idx = self._u16()
            if idx >= len(self.globals):
                raise Trap("T07", self.op_pc)
            self.push(self.globals[idx])
            return None
        if op == 0x43:  # gstore
            idx = self._u16()
            if idx >= len(self.globals):
                raise Trap("T07", self.op_pc)
            self.globals[idx] = self.pop()
            return None
        if op == 0x44:  # dload
            addr = self.pop()
            if addr < 0 or addr >= len(self.data):
                raise Trap("T08", self.op_pc)
            self.push(self.data[addr])
            return None
        if op == 0x45:  # gloadx
            idx = self.pop()
            if idx < 0 or idx >= len(self.globals):
                raise Trap("T07", self.op_pc)
            self.push(self.globals[idx])
            return None
        if op == 0x46:  # gstorex
            idx = self.pop()
            value = self.pop()
            if idx < 0 or idx >= len(self.globals):
                raise Trap("T07", self.op_pc)
            self.globals[idx] = value
            return None

        # -- 4.6 output
        if op == 0x50:  # print
            self.stdout.write(ops.format_decimal(self.pop()))
            return None
        if op == 0x51:  # printc
            self.stdout.write(bytes((self.pop() & 0xFF,)))
            return None
        if op == 0x52:  # prints
            length = self.pop()
            addr = self.pop()
            if length < 0 or addr < 0 or addr + length > len(self.data):
                raise Trap("T08", self.op_pc)
            self.stdout.write(self.data[addr : addr + length])
            return None

        # -- 4.7 diagnostics
        if op == 0x60:  # trap
            code = self._u8()
            raise Trap("T10", self.op_pc, f"trap instruction (code={code})")

        raise Trap("T05", self.op_pc)

    def _arith(self, op: int) -> None:
        if op == 0x15:  # neg
            self.push(ops.neg(self.pop()))
            return None
        if op == 0x19:  # not
            self.push(ops.bit_not(self.pop()))
            return None

        b, a = self.pop(), self.pop()
        if op in (0x13, 0x14) and b == 0:
            raise Trap("T04", self.op_pc)
        self.push(_BINARY[op](a, b))
        return None

    def _call(self) -> None:
        target = self._u32()
        nargs = self._u8()
        nlocals = self._u8()

        if nargs > nlocals:
            raise Trap("T06", self.op_pc)
        if len(self.frames) >= isa.MAX_CALL_DEPTH:
            raise Trap("T03", self.op_pc)

        # Pop into locals back-to-front so the first value pushed lands in locals[0].
        slots = [0] * nlocals
        for i in range(nargs - 1, -1, -1):
            slots[i] = self.pop()

        if target >= self.code_len:
            raise Trap("T09", self.op_pc)

        self.frames.append(Frame(slots, self.pc))
        self.pc = target
        return None

    def _ret(self) -> int | None:
        value = self.pop()
        finished = self.frames.pop()
        if not self.frames:
            # Returning from the entry frame terminates the program (spec section 5.3).
            return value & 0xFF
        self.pc = finished.return_pc
        self.push(value)
        return None


_BINARY = {
    0x10: ops.add,
    0x11: ops.sub,
    0x12: ops.mul,
    0x13: ops.div,
    0x14: ops.mod,
    0x16: ops.bit_and,
    0x17: ops.bit_or,
    0x18: ops.bit_xor,
    0x1A: ops.shl,
    0x1B: ops.shr,
    0x1C: ops.ushr,
}

_COMPARE = {
    0x20: lambda a, b: a == b,
    0x21: lambda a, b: a != b,
    0x22: lambda a, b: a < b,
    0x23: lambda a, b: a <= b,
    0x24: lambda a, b: a > b,
    0x25: lambda a, b: a >= b,
}


def execute(module: Module, stdout: BinaryIO, stderr: BinaryIO) -> int:
    """Run ``module``, returning the process exit status.

    Handles the trap protocol: prior stdout is flushed first, then the single trap line
    goes to stderr, then the status is 70.
    """
    machine = VM(module, stdout)
    try:
        status = machine.run()
    except Trap as trap:
        stdout.flush()
        stderr.write(trap.line())
        stderr.flush()
        return isa.EXIT_TRAP
    except RecursionError:  # pragma: no cover - defensive; the VM loop is iterative
        stdout.flush()
        stderr.write(isa.trap_line("T03", machine.op_pc))
        stderr.flush()
        return isa.EXIT_TRAP
    stdout.flush()
    return status
